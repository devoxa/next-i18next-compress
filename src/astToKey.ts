import * as BabelTypes from '@babel/types'
import colors from 'colors/safe'

export type AbstractSyntaxTree = Array<
  | BabelTypes.Expression
  | BabelTypes.SpreadElement
  | BabelTypes.JSXNamespacedName
  | BabelTypes.ArgumentPlaceholder
  | BabelTypes.TemplateElement
  | BabelTypes.TSType
  | BabelTypes.JSXText
  | BabelTypes.JSXExpressionContainer
  | BabelTypes.JSXSpreadChild
  | BabelTypes.JSXElement
  | BabelTypes.JSXFragment
  | BabelTypes.JSXEmptyExpression
>

export interface AstToKeyOptions {
  code: string
  jsx?: boolean
}

export function astToKey(ast: AbstractSyntaxTree, options: AstToKeyOptions): string {
  let key = ''

  // Ignore empty JSXText nodes, because they get stripped away by React
  // and interfere with getting the correct indexes for the JSXElement nodes.
  ast = ast.filter((astNode) => {
    if (astNode.type === 'JSXText' && astNode.value.trim() === '') return false
    return true
  })

  for (let i = 0; i !== ast.length; i++) {
    const astNode = ast[i]

    // This handles a normal string literal, like 'Foobar'
    if (astNode.type === 'StringLiteral') {
      key += astNode.value
      continue
    }

    // This handles a template string, like `Foobar`. It it build out of "quasis", which
    // are the template elements (= string literals), and expressions. We handle them all-in-one.
    if (astNode.type === 'TemplateLiteral') {
      key += astToKey([...astNode.expressions, ...astNode.quasis], options)
      continue
    }

    // This is a template element (= string literal) of a template string (inside of quasis).
    if (astNode.type === 'TemplateElement') {
      key += astNode.value.raw
      continue
    }

    // This is text inside of a JSX component. It may have whitespace, which is ignored
    // by React, so we have to clean it up to match what is expected.
    if (astNode.type === 'JSXText') {
      let value = astNode.value

      value = value.replace(/\n *$/g, '') // Ignore trailing newlines
      value = value.replace(/\n */g, ' ') // Turn remaining newlines into spaces
      if (i === 0) value = value.trimStart() // Trim the start if this is the first node
      if (i === ast.length - 1) value = value.trimEnd() // Trim the end if this is the last node

      key += value
      continue
    }

    // This is a JSX component, like `<Trans><Foo>...</Foo></Trans>`
    if (options.jsx && astNode.type === 'JSXElement') {
      const childNodesKey = astToKey(astNode.children, options)

      key += `<${i}>${childNodesKey}</${i}>`
      continue
    }

    // This is an interpolated expression, like `<Trans>Foo {...}</Trans>`
    if (astNode.type === 'JSXExpressionContainer') {
      key += astToKey([astNode.expression], options)
      continue
    }

    // This is an interpolated comment, like `<Trans>Foo {/* bar */}</Trans>`. We ignore these.
    if (astNode.type === 'JSXEmptyExpression') {
      continue
    }

    // This is an interpolated variable, like "t(`<Trans>Foo {{bar}}</Trans>`)" in JSX components.
    // "t('Foo {{bar}}')" in template strings is handled like a string literal.
    // (!) Keep in mind that `{{ name }}` is different than `{{name}}`, so we handle it exactly as-is!
    if (options.jsx && astNode.type === 'ObjectExpression') {
      // istanbul ignore next
      if (!astNode.start || !astNode.end) {
        throw new Error('Start or end of a AST node are missing, please file a bug report!')
      }

      // We slice the code out of the file instead of trying to recreate it from the AST
      // because I value my continued sanity.
      const astNodeCode = options.code.slice(astNode.start, astNode.end)

      key += `{${astNodeCode}}`
      continue
    }

    // We deliberately do not handle the following types:
    // - JSXSpreadChild is not supported by React/NextJS
    // - CallExpression is not supported `i18next`
    // - JSXFragment is not supported by `i18next`
    // - Identifier is generally a misuse (should be `{{variable}}` instead of `{variable}`)
    // - MemberExpression is generally a misuse (not supported by React)

    throw new UnsupportedAstTypeError(astNode, options.code)
  }

  return key
}

export class UnsupportedAstTypeError extends Error {
  constructor(astNode: { type: string; start: number | null; end: number | null }, code: string) {
    let message =
      `[next-i18next-compress] Unsupported AST type: ` +
      `We do not know how to handle "${astNode.type}"`

    if (code && astNode.start && astNode.end) {
      const codeRange = printCodeRange(code, astNode.start, astNode.end, 30)
      message += ` in this part of your code:\n${codeRange}`
    } else {
      message += '.'
    }

    super(message)
  }
}

function printCodeRange(code: string, start: number, end: number, padding: number) {
  const slicedStartPadding = code.slice(Math.max(0, start - padding), start)
  const slicedCode = code.slice(start, end)
  const slicedEndPadding = code.slice(end, Math.min(code.length, end + padding))

  // istanbul ignore next
  const format = process.env.NODE_ENV === 'test' ? (x: string) => x : colors.red

  return slicedStartPadding + format(slicedCode) + slicedEndPadding
}
