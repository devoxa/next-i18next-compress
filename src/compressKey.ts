import crypto from 'crypto'

export function compressKey(string: string, hashLength: number) {
  // Don't compress keys with interpolated React components
  if (string.match(/<.*?>/g)) return string

  // Strip whitespace to make keys extracted from Babel consistent with keys in locale files
  string = string.trim().replace(/( |\n)+/g, ' ')

  // Hash the key's content into 6 characters. A collision may occur with a sufficiently
  // large locale file. Using 6 characters and a locale file with 100 keys, the probability of a
  // collision is about 0.029%.
  return crypto.createHash('sha256').update(string).digest('hex').slice(0, hashLength)
}
