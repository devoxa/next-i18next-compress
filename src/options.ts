export interface Options {
  hashLength?: number
}

export function mergeDefaultOptions(pOptions: Partial<Options> = {}) {
  if (pOptions.hashLength && (pOptions.hashLength < 4 || pOptions.hashLength > 64)) {
    const message =
      `[next-i18next-compress] Misconfiguration detected: ` +
      `The "hashLength" option has to be between 4 and 64`

    throw new Error(message)
  }

  return {
    hashLength: pOptions.hashLength || 6,
  }
}
