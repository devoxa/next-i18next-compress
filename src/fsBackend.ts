import { compressKey } from './compressKey'

function parseLocaleFile(string: string) {
  const json = JSON.parse(string)

  const compressedKeys: Record<string, string> = {}
  const compressedJson: Record<string, string> = {}

  Object.keys(json).forEach((key) => {
    const compressedKey = compressKey(key, 6)

    // The compressed key already exists, throw a collision error
    if (compressedKeys[compressedKey]) {
      const message =
        `[next-i18next-compress] Compression collision detected: ` +
        `"${compressedKeys[compressedKey]}" and "${key}" compress to the same hash "${compressedKey}"`

      throw new Error(message)
    }

    compressedKeys[compressedKey] = key
    compressedJson[compressedKey] = json[key]
  })

  return compressedJson
}

export default { backend: { parse: parseLocaleFile } }
