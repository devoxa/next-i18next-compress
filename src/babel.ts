import * as BabelTypes from '@babel/types'
import { Visitor } from '@babel/traverse'
import { compressKey } from './compressKey'
import { mergeDefaultOptions, Options } from './options'

// We keep a set of processed nodes, because Babel may traverse the same node twice,
// which would cause us to compress the key twice.
const processedNodes = new Set()

export interface Babel {
  types: typeof BabelTypes
}

type State = { opts?: Partial<Options>; file: { opts: { filename?: string } } }

export default function nextI18nextCompressBabelPlugin(
  babel: Babel
): { name: string; visitor: Visitor } {
  const t = babel.types

  return {
    name: 'next-i18next-compress',

    visitor: {
      CallExpression(path, _state) {
        const state = _state as State

        // Do not process any files in development
        if (process.env.NODE_ENV === 'development') {
          return
        }

        // Do not process any files in `node_modules/`, since this can cause issues with minified code
        if (state.file.opts.filename && state.file.opts.filename.includes('/node_modules/')) {
          return
        }

        const options = mergeDefaultOptions(state.opts)

        // istanbul ignore next
        if (processedNodes.has(path.node)) return

        // Only handle functions with the name `t` and at least one argument
        if (!t.isIdentifier(path.node.callee, { name: 't' })) return
        if (path.node.arguments.length === 0) return

        // We don't support cases where the argument is not a string, because
        // we can't figure out what the actual value is easily.
        if (path.node.arguments[0].type !== 'StringLiteral') {
          return unsupportedCodeUse('`t(variable)` is not supported, use a string literal instead.')
        }

        // Compress the argument value (this is either the `i18nKey` or the natural key)
        path.node.arguments[0].value = compressKey(path.node.arguments[0].value, options.hashLength)

        processedNodes.add(path.node)
      },

      JSXElement(path, _state) {
        const state = _state as State

        // Do not process any files in development
        if (process.env.NODE_ENV === 'development') {
          return
        }

        // Do not process any files in `node_modules/`, since this can cause issues with minified code
        if (state.file.opts.filename && state.file.opts.filename.includes('/node_modules/')) {
          return
        }

        const options = mergeDefaultOptions(state.opts)

        // istanbul ignore next
        if (processedNodes.has(path.node)) return

        // Only handle JSX elements with the name `Trans`
        if (!t.isJSXIdentifier(path.node.openingElement.name, { name: 'Trans' })) return

        // We don't support cases where a variable is spread into the attributes,
        // because there might be a `i18nKey` in it that we might overwrite.
        const elementMixedAttributes = path.node.openingElement.attributes
        if (elementMixedAttributes.some((x) => x.type === 'JSXSpreadAttribute')) {
          return unsupportedCodeUse(
            '`<Trans {...variable}>` is not supported, use explicit attributes instead.'
          )
        }
        const elementJsxAttributes = elementMixedAttributes as Array<BabelTypes.JSXAttribute>

        // Get the content of the `i18nKey` attribute, if it exists and has a value
        let i18nKeyAttributeValue
        const i18nKeyAttribute = elementJsxAttributes.find((x) => x.name.name === 'i18nKey')
        if (i18nKeyAttribute && i18nKeyAttribute.value) {
          // We don't support cases where the attribute is not a string, because
          // we can't figure out what the actual value is easily.
          if (i18nKeyAttribute.value.type !== 'StringLiteral') {
            return unsupportedCodeUse(
              '`<Trans i18nKey={variable}>` is not supported, use a string literal instead.'
            )
          }

          i18nKeyAttributeValue = i18nKeyAttribute.value.value
        }

        // Get the key based on the children, if they exist
        let childrenKey
        if (path.node.children.length > 0) {
          childrenKey = childrenToKey(path.node.children as BabelTypes.JSXElement['children'])
        }

        // The key is either the `i18nKey` attribute or the child text node
        const key = (i18nKeyAttributeValue || childrenKey) as string

        // Generate the new `i18nKey` attribute with the compressed key
        const keyAttribute = {
          type: 'JSXAttribute',
          name: { type: 'JSXIdentifier', name: 'i18nKey' },
          value: { type: 'StringLiteral', value: compressKey(key, options.hashLength) },
        }

        // Strip the element of any existing `i18nKey` attribute and insert the new one
        path.node.openingElement.attributes = path.node.openingElement.attributes.filter(
          (x) => x.type === 'JSXAttribute' && x.name.name !== 'i18nKey'
        )
        path.node.openingElement.attributes.push(keyAttribute as any) // eslint-disable-line @typescript-eslint/no-explicit-any

        const canTurnIntoSelfClosing =
          path.node.children.length === 0 ||
          (path.node.children.length === 1 && path.node.children[0].type === 'JSXText')

        if (canTurnIntoSelfClosing) {
          // If there is no children or only one text child, turn it into a self-closing element
          // with no children or additional closing element.
          // `<Trans>Foo</Trans>` -> `<Trans i18nKey={...} />`
          path.node.children = []
          delete path.node.closingElement
          path.node.openingElement.selfClosing = true
        } else {
          // If there is something else, at least compress the text nodes to single characters
          // since they will be replaced by the translated text.
          // `<Trans>Foo <Link>Bar</Link></Trans>` -> `<Trans i18nKey={...}>~<Link>~</Link></Trans>`
          compressChildTextNodes(path.node.children as BabelTypes.JSXElement['children'])
        }

        processedNodes.add(path.node)
      },
    },
  }
}

function unsupportedCodeUse(message: string) {
  throw new Error('[next-i18next-compress] Unsupported code use: ' + message)
}

export function childrenToKey(children: BabelTypes.JSXElement['children']): string {
  let key = ''

  // We ignore empty text nodes since they get stripped by React
  children = children.filter((child) => {
    if (child.type === 'JSXText' && child.value.trim() === '') return false
    return true
  })

  for (let i = 0; i !== children.length; i++) {
    const child = children[i]

    if (child.type === 'JSXElement') {
      key += `<${i}>` + childrenToKey(child.children) + `</${i}>`
      continue
    }

    if (child.type === 'JSXText') {
      let text = child.value

      // Ignore trailing newlines
      text = text.replace(/\n *$/g, '')

      // Turn remaining newlines into spaces
      text = text.replace(/\n */g, ' ')

      // Trim the start if we are the first child
      if (i === 0) {
        text = text.trimStart()
      }

      // Trim the end if we are the last child
      if (i === children.length - 1) {
        text = text.trimEnd()
      }

      key += text
      continue
    }

    if (child.type === 'JSXExpressionContainer') {
      // Take the value of string literal expressions like `{'  '}` as-is
      if (child.expression.type === 'StringLiteral') {
        key += child.expression.value
        continue
      }

      // Add markers for interpolated variables by name
      if (child.expression.type === 'Identifier') {
        key += `{${child.expression.name}}`
        continue
      }

      // Ignore comments
      if (child.expression.type === 'JSXEmptyExpression') {
        continue
      }

      // We don't want to handle interpolated expressions. i18next-parser does, but it's a bit iffy
      // to do (they take the entire source and then slice the expression out).
      throw new Error('[next-i18next-compress] Unknown AST type: ' + child.expression.type)
    }

    if (child.type === 'JSXSpreadChild') {
      unsupportedCodeUse('`<Trans>{...variable}</Trans>` is not supported (by NextJS)')
    }

    if (child.type === 'JSXFragment') {
      unsupportedCodeUse('`<Trans>Text <>here</></Trans>` is not supported')
    }

    throw new Error('[next-i18next-compress] Unknown AST type: ' + child.type)
  }

  return key
}

function compressChildTextNodes(children: BabelTypes.JSXElement['children']): void {
  for (let i = 0; i !== children.length; i++) {
    const child = children[i]

    if (child.type === 'JSXElement') {
      compressChildTextNodes(child.children)
      continue
    }

    // We ignore empty text nodes since they get stripped by React
    if (child.type === 'JSXText' && child.value.trim() !== '') {
      child.value = '~'
      continue
    }
  }
}
