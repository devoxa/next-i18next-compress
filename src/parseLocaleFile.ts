import { compressKey } from './compressKey'
import { mergeDefaultOptions, Options } from './options'

export function parseLocaleFile(string: string, pOptions?: Options) {
  const options = mergeDefaultOptions(pOptions)
  const json = JSON.parse(string)

  if (process.env.NODE_ENV === 'development') {
    return json
  }

  const compressedKeys: Record<string, string> = {}
  const compressedJson: Record<string, string> = {}

  Object.keys(json).forEach((key) => {
    const compressedKey = compressKey(key, options.hashLength)

    // The compressed key already exists, throw a collision error
    if (compressedKeys[compressedKey]) {
      const message =
        `[next-i18next-compress] Compression collision: ` +
        `"${compressedKeys[compressedKey]}" and "${key}" compress to the same hash "${compressedKey}".` +
        `Try increasing the "hashLength" option or splitting your locale file into more namespaces.`

      throw new Error(message)
    }

    compressedKeys[compressedKey] = key
    compressedJson[compressedKey] = json[key]
  })

  return compressedJson
}
