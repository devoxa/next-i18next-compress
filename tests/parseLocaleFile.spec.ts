import { parseLocaleFile } from '../src/parseLocaleFile'
import * as compressKeyModule from '../src/compressKey'

const compressKeySpy = jest.spyOn(compressKeyModule, 'compressKey')

const localeFileString = JSON.stringify({
  'Email address': 'E-Mail Adresse',
  'Forgot password': 'Passwort vergessen',
  'Or <1>start your 30-day free trial</1>':
    'Oder <1>beginne deine 30-Tage kostenlose Probephase</1>',
  'Your sign in credentials were incorrect.': 'Deine Anmeldedaten waren inkorrekt.',
  'Happy birthday, {{name}}!': 'Alles Gute zum Geburtstag, {{name}}!',
})

describe('parseLocaleFile', () => {
  it('correctly compresses the keys of the locale file', () => {
    expect(parseLocaleFile(localeFileString)).toMatchSnapshot()
  })

  it('can configure the length of the compressed key', () => {
    expect(parseLocaleFile(localeFileString, { hashLength: 16 })).toMatchSnapshot()
  })

  it('throws an error if there are any compression collisions', async () => {
    compressKeySpy.mockImplementation(() => 'foobar')
    expect(() => parseLocaleFile(localeFileString)).toThrowErrorMatchingSnapshot()
  })

  it('does nothing if running in development', () => {
    process.env.NODE_ENV = 'development'
    expect(parseLocaleFile(localeFileString)).toEqual(JSON.parse(localeFileString))
  })
})
