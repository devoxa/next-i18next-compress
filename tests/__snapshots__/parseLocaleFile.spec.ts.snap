// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseLocaleFile can configure the length of the compressed key 1`] = `
{
  "0d1f217dd077602b": "Oder <1>beginne deine 30-Tage kostenlose Probephase</1>",
  "bb42f8cd525dba08": "Passwort vergessen",
  "d81b0fdda384fc7e": "Alles Gute zum Geburtstag, {{name}}!",
  "d9055ff096c25633": "Deine Anmeldedaten waren inkorrekt.",
  "f2488fd4ef4adbc6": "E-Mail Adresse",
}
`;

exports[`parseLocaleFile correctly compresses the keys of the locale file 1`] = `
{
  "0d1f21": "Oder <1>beginne deine 30-Tage kostenlose Probephase</1>",
  "bb42f8": "Passwort vergessen",
  "d81b0f": "Alles Gute zum Geburtstag, {{name}}!",
  "d9055f": "Deine Anmeldedaten waren inkorrekt.",
  "f2488f": "E-Mail Adresse",
}
`;

exports[`parseLocaleFile throws an error if there are any compression collisions 1`] = `"[next-i18next-compress] Compression collision: "Email address" and "Forgot password" compress to the same hash "foobar".Try increasing the "hashLength" option or splitting your locale file into multiple namespaces."`;
